torch/distributed/_tensor/experimental/attention.py

```diff
@@ -0,0 +1,467 @@
+import contextlib
+from enum import Enum
+from typing import Dict, Generator, List, Optional, Tuple, Union
+
+import torch
+import torch.distributed as dist
+import torch.distributed._functional_collectives as ft_c
+from torch import nn
+from torch.distributed._tensor import distribute_module, DTensor, Replicate
+from torch.distributed.device_mesh import DeviceMesh
+from torch.distributed.tensor.parallel.style import ParallelStyle
+
+aten = torch.ops.aten
+
+
+def sdpa_handler(
+    op_call: torch._ops.OpOverload,
+    args: Tuple[object, ...],
+    kwargs: Dict[str, object],
+) -> object:
+    # extract local tensor and sharding infos to a OpInfo
+    op_info = DTensor._op_dispatcher.unwrap_to_op_info(op_call, args, kwargs)
+
+    # sharding propagation
+    DTensor._op_dispatcher.sharding_propagator.propagate(op_info)
+    output_sharding = op_info.output_sharding
+    assert output_sharding is not None, "output sharding should not be None"
+    assert not output_sharding.needs_redistribute, "inputs need to be redistributed"
+
+    local_results = _scaled_dot_product_flash_attention(
+        op_info.mesh,
+        *op_info.local_args,  # type: ignore[arg-type]
+        **op_info.local_kwargs,  # type: ignore[arg-type]
+    )
+
+    return DTensor._op_dispatcher.wrap(local_results, output_sharding.output_spec)
+
+
+def _merge_sdpa(
+    chunks: List[torch.Tensor], logsumexps: List[torch.Tensor]
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    """
+    This merges multiple scaled dot product attention chunks by using the
+    provided logsumexps to rescale the chunks before summing.
+
+    Args:
+        chunks (List[torch.Tensor]): A list of scaled dot product attention chunks
+        logsumexps (List[torch.Tensor]): A list of logsumexps for each chunk
+
+    Returns:
+        out (torch.Tensor): The merged scaled dot product attention
+        softmax_lse (torch.Tensor): The logsumexp of the merged scaled dot product attention
+    """
+    assert len(chunks) == len(logsumexps)
+
+    softmax_lse = torch.stack([lse.exp() for lse in logsumexps]).sum(dim=0).log_()
+
+    out = []
+    for i, (chunk, chunk_lse) in enumerate(zip(chunks, logsumexps)):
+        softmax_lse_corrected = torch.exp(chunk_lse - softmax_lse)
+        out_corrected = chunk * softmax_lse_corrected.unsqueeze(-1).to(chunk.dtype)
+        out.append(out_corrected)
+    out = torch.stack(out).sum(dim=0)
+
+    return out, softmax_lse
+
+
+def _scaled_dot_product_flash_attention(
```

Comment:
nit on naming, to distinguish with sdpa itself: `_scaled_dot_product_flash_ring_attention`

Reply:
done